Prototype of the RxJS interop layer for signals

alxhub opened the pull request:
### For context, see this [Github Discussion](https://github.com/angular/angular/discussions/49090) on our experiments with fine-grained reactivity in Angular

(the first commit here is #49150 which serves as a base for this implementation)

This is the prototype implementation of the RxJS interop layer, sketched out as a separate package (`@angular/rxjs-interop`).

alxhub committed:
feat(core): prototype implementation of @angular/core/rxjs-interop

This commit adds the basic sketch for the implementation of `fromObservable`
and `fromSignal`, the two basic primitives which form the RxJS interop layer
with signals.

tomalaforge commented on packages/rxjs-interop/src/from_observable.ts:
```diff
@@ -0,0 +1,90 @@
+  obs$.subscribe({
+    next: value => state.set({kind: StateKind.Value, value}),
+    error: error => state.set({kind: StateKind.Error, error}),
+  });
```
Question: where do you unsubscribe from the observable ? 😮

JeanMeche replied:
There is a todo about it 2 lines below 😉

pkozlowski-opensource replied:
missing piece of puzzle: https://github.com/angular/angular/pull/49158

ajafff replied:
@pkozlowski-opensource why would you use a callback of Angular's lifecycle to unsubscribe?
the subscription is made when the signal is created, so unsubscription should only happen if the signal is garbage collected (using WeakRef and/or FinalizationRegistry)

on the other hand, if the subscription was lazy on first access of the signal's value, the unsubscription should happen when there are no more consumers. but that would probably not fit the design goals of signals (e.g. no side effects)

pkozlowski-opensource replied:
> on the other hand, if the subscription was lazy on first access of the signal's value, the unsubscription should happen when there are no more consumers. but that would probably not fit the design goals of signals (e.g. no side effects)

Agreed, our intention is to have eager subscription, precisely for the reason of avoiding any side-effects in signal reads.

kbrilla commented on packages/rxjs-interop/src/from_observable.ts:
```diff
@@ -0,0 +1,90 @@
+    const current = state();
+    switch (current.kind) {
+      case StateKind.NoValue:
+        throw new Error(`fromObservable() signal read before the Observable emitted`);
```
Two things here:
What about info that observable completed? 
Why throw an error when subscribing to observable without a value yet?

Both states are usefull to known. As proof of that, please have a look at [ngrx let pipe](https://ngrx.io/guide/component/let#using-suspense-template)

alxhub replied:
> What about info that observable completed?

Signals don't have a notion of "completed" or "that was the last value and this will never update again", so there's no need to track the completed status of the Observable.

> Why throw an error when subscribing to observable without a value yet?

The error isn't thrown if you subscribe, only if you try to read the value before the Observable has produced one yet (and only if there is no initial value provided).

Signals are synchronous, and should always have a current value. This isn't true of Observables, which aren't guaranteed to produce a value (maybe ever). To bridge the two, there are two strategies:

1. Return a "default" or initial value until the Observable actually emits. This is what `async` pipe does with `null`, and what `fromObservable` does if you actually provide its second argument.

2. Throw if someone tries to read the value before it's ready. This is useful when you know you're dealing with an Observable which _will_ produce a value fast enough that you don't need to wait for it.

kbrilla replied:
Sure signals have no error state but observables have few states that are useful to keep like 

1. if there was an value
2. if there was an error
3. if the observable completed
4. if we got an value

All of that is useful for example with http requests where you want to know that you are waiting for http request, when you get and value or error.  
With throwing error when You read a value when in `NoValue` state You lose information if there is no value yet or observable completed without error and value.
The developer can use all those information so they should not be dropped when converting to signal. 

There could be returned value like 
```
type SignalFromObservable<T> = {type: 'NoValue', completed: boolean} | {type: 'Value', value: T, completed: boolean} | {type: 'Error', error: any} | {};
fromObservable<T>(observable: Observable<T>): Signal<SignalFromObservable<T>>
```
there is also an additional benefit as we do not add anything to the type of value T as opposed to T | U.
now we can show all info about the request
```html
<ng-container *ngIf="fromObsSignal() as signal">
<span *ngIf="signal.type === 'NoValue'">
There is no value and the observable {{ signal.completed ? 'did complete!' : 'did not complete'}}  
</span>
<span *ngIf="signal.type === 'Value'">
Last value was {{signal.value}} and the observable {{ signal.completed ? 'did complete!' : 'did not complete'}}
</span>
<span *ngIf="signal.type === 'Error'">
Observable throw and error {{signal.error}}!
</span>
</ng-container>
```

```html
<ng-container *ngIf="request() as request">
<span *ngIf="request.type === 'NoValue'">
Loading...
</span>
<span *ngIf="request.type === 'Value'">
{{request.value}}
</span>
<span *ngIf="signal.type === 'Error'">
There was an error!
</span>
</ng-container>
```

I think that it would be beneficial to also have this kind of info - maybe in another helper - there does not need to be just one.

pkozlowski-opensource replied:
I hear your points but the utility being developed here is _not_ meant to plug _any_ observable. Rather, focus on observables that represent values over time and match conceptual model of signals.

More specifically, we would _never_ recommend modelling HTTP interactions as signals.

muhamedkarajic replied:
> * Return a "default" or initial value until the Observable actually emits. This is what `async` pipe does with `null`, and what `fromObservable` does if you actually provide its second argument.
> * Throw if someone tries to read the value before it's ready. This is useful when you know you're dealing with an Observable which _will_ produce a value fast enough that you don't need to wait for it.

You can have a third strategy which is loading, this means its not available. Could also be a LoadingSignal type. Again just my 2 cents :)

GabrielBB commented on goldens/public-api/core/index.md:
```diff
@@ -1421,6 +1445,12 @@ export interface TypeDecorator {
 }
 
+// @public
+export function untrack<T>(nonReactiveReadsFn: () => T): T;
```
I thought this was [renamed](https://github.com/angular/angular/pull/49091#discussion_r1113436592)

alxhub replied:
Fixed!

ajafff commented on packages/rxjs-interop/src/from_signal.ts:
```diff
@@ -0,0 +1,44 @@
+  // Pipe via `shareReplay` to share the single backing `effect` across all interested subscribers.
+  // This of course has timing implications for when new subscribers show up. We turn on `refCount`
+  // so that once all subscribers are unsubscribed, the underlying `effect` can be cleaned up.
+  return signal$.pipe(shareReplay({refCount: true}));
```
you need to add `bufferSize: 1`, otherwise `shareReplay` will replay all emissions instead of just the most recent one

alxhub replied:
Good catch! Added, and added a test.

GabrielBB commented on packages/rxjs-interop/src/from_signal.ts:
```diff
@@ -0,0 +1,44 @@
+ * begin to propagate to subscribers with the same timing as an `effect` watching that value. New
+ * subscribers beyond the first will receive the current value immediately.
+ *
+ * When no more subscribers are subscribed, the internal `effect` will be cleaned up.
```
```suggestion
 * When no more subscriptions, the internal `effect` will be cleaned up.
```

jelbourn replied:
Active voice here would be "When this observable has no subscriptions"

GabrielBB commented on goldens/public-api/rxjs-interop/index.md:
```diff
@@ -0,0 +1,21 @@
+export function fromObservable<T, U>(obs$: Observable<T>, initialValue: U): Signal<T | U>;
+
+// @public
+export function fromSignal<T>(source: Signal<T>): Observable<T>;
```
I wonder what is better: Mention what is received VS Mention what is returned. `fromObservable` / `fromSignal`  `toObservable` / `toSignal`

phhien203 replied:
I think it would make it consistent with API from RxJS, e.g. `fromEvent`, `fromFetch`

GabrielBB replied:
Ah true

pkozlowski-opensource replied:
Yes, let's keep it consistent with the RxJS naming.

jelbourn commented on packages/rxjs-interop/src/from_observable.ts:
```diff
@@ -0,0 +1,90 @@
+ *
+ * @developerPreview
+ */
+export function fromObservable<T, U>(obs$: Observable<T>, initialValue: U): Signal<T|U>;
```
Do we currently use the `$` suffix pattern anywhere else in the Angular public APIs? If not, I would prefer to remain consistent.

alxhub replied:
Removed :)

jelbourn commented on packages/rxjs-interop/src/from_signal.ts:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Exposes the value of an Angular `Signal` as an RxJS `Observable`.
+ *
+ * The returned `Observable` has `shareReplay` semantics. When subscribed, the current value will
+ * begin to propagate to subscribers with the same timing as an `effect` watching that value. New
```
```suggestion
 * The returned `Observable` has `shareReplay` semantics. When subscribed, the current value
 * starts propagating to subscribers with the same timing as an `effect` watching that value. New
```
nit: present tense

jelbourn reviewed (approved):
LGTM barring open comments

jnizet commented on packages/rxjs-interop/test/from_signal_spec.ts:
```diff
@@ -0,0 +1,150 @@
+
+  it('should produce an observable that tracks a signal', async () => {
+    const counter = signal(0);
+    const counter$ = fromSignal(counter).pipe(take(3), toArray()).toPromise();
```
A bit unusual to use the $ suffix on a promise. Besides, toPromise() is deprecated in RxJS7.

alxhub replied:
Yep, good catch - fixed.

Angular's on RxJS 6 though, so no fancy `latestValueFrom` for us.

tomalaforge replied:
What stoping you from moving to a higher version of Rxjs?

alxhub replied:
Google uses RxJS 6 internally, so Angular needs to build with it.

We could (and probably will) upgrade to using 7, but we'd still need to be _compatible_ with 6, so it's not been a priority.

elgreco247 commented on packages/rxjs-interop/src/from_observable.ts:
```diff
@@ -0,0 +1,90 @@
+
+  source.subscribe({
+    next: value => state.set({kind: StateKind.Value, value}),
+    error: error => state.set({kind: StateKind.Error, error}),
```
Maybe check if error (the parameter) is an instance of Error and if not wrap it in a new instance of Error before setting the state signal? Otherwise you can't tell if what `fromObservable` is returning is a value or an error.

elgreco247 replied:
Hmmm... so that means it would not be advisable to use a signal created by `fromObservable` directly in a template, like here: https://github.com/angular/angular/discussions/49090#discussioncomment-5109124 .

I'd need something like this for my template:

```javascript
signal = fromObservable(data$);
signalError = computed(() => {
  try {
    signal();
    return;
  } catch(error) {
    return error;
  }
});
```

pkozlowski-opensource replied:
Well, as soon as an observable emits an error I don't think we can "recover" and in this case throwing an error seems like the appropriate thing to do.

GabrielBB replied:
I wonder if Angular should include a  signalutility function to recover from errors

GabrielBB replied:
`const newSignal = signal.onError(err => ({...}))`

alxhub replied:
Or from a `computed` as @elgreco247 shows - it could process the error and return whatever state for the UI that it wants.

elgreco247 replied:
@pkozlowski-opensource agreed. I was just wondering what would happen if I used a signal created with `fromObservable` in an `*ngIf`; if an error is thrown, then there is no way to catch it in the template. Hence the need for the `computed`.

pkozlowski-opensource replied:
Yes, I believe that existing signal APIs have enough primitives to build error recovery if needed. To be honest, it doesn't even need to be `computed` - it is enough to do:

```
const obsSignal = fromObservable(obs$);
const signalWithErrRecovery = () => {
  try {
    return signal();
  } catch(error) {
    return error;
  }
});
```
Let's start _without_ this utility and if the demand is high / there are many practical use-cases, we can always add it as its trivial.

chen-yangfan commented:
can I remove zone when I use Signal？

josephperrott reviewed (approved):
LGTM for dev-infra

Note: this will require us to create this package on npm prior to being able to perform a release after this pull request lands.

@alxhub I also wanted to ask, why is this not in `core` as well? It seems to me that it makes more sense to have this colocated with signals rather than ths which feels a little like an indirection.

menosprezzi commented:
Hey @alxhub! Nice work! But why dont just implement the [InteropObservable](https://github.com/ReactiveX/rxjs/blob/f93e5c86a37af3b553a950b73eb346ee2806eddb/src/internal/types.ts#L97) type to build the bridge between rxjs and signals? Is there a reason?
If we did so, translating between both will ease!

Think how the usage sounds better:

```ts
combineLatest([myPrimitiveSignal, myObservable$]) // ...and so on!
```

pkozlowski-opensource commented on packages/rxjs-interop/README.md:
```diff
@@ -0,0 +1,4 @@
+Angular RxJS Interop
+=========
+
+This package includes utilities related to using the RxJS library in conjunction with Angular's signal-based reactivity system.
```
Do we want to mention signals / reactivity explicitly here? I guess long-term we want to put other non-signal, RxJS related stuff here.

pkozlowski-opensource commented on packages/rxjs-interop/package.json:
```diff
@@ -0,0 +1,24 @@
+  "peerDependencies": {
+    "@angular/core": "0.0.0-PLACEHOLDER",
+    "rxjs": "^6.5.3 || ^7.4.0",
+    "zone.js": "~0.11.4 || ~0.12.0"
```
Do we need explicit peer dep on zone.js?

csaszika replied:
@alxhub Can you explain please this part please?
I mean it was a `rxjs-interop` package, probably the plan was `@angular/rxjs-interop`.
Now as I see, `rxjs-interop` will be the part of `core`, so `zone.js` is still will be a dependency there. As I heard, `signals` are advertised as zoneless solutions, which is currently the only (main) reason for me why we should use them. 
I would like to understand that this pull request is just a step forward or there are other reason why we still need `zone.js`.

I think having another npm package as `@angular/rxjs-interop` or something is better until `zone.js` cannot be eliminated from `core`. In long term I agree, it should be in  `core` library.

pkozlowski-opensource commented on packages/rxjs-interop/src/from_observable.ts:
```diff
@@ -0,0 +1,90 @@
+ *
+ * @developerPreview
+ */
+export function fromObservable<T, U>(source: Observable<T>, initialValue: U): Signal<T|U>;
```
I find it surprising that the type of the signal and initial state are different. What is the use-case that we are targeting here?

jnizet replied:
My guess is that it's mainly useful when you have an `Observable<Something>` and you want `null` or `undefined` as default value.

kbrilla replied:
implementation of rxjs `firstValueFrom` and `lastValueFrom` also have different type for defaultValue. 
But maybe instead of `initialValue` rxjs `defaultValue` is more fitting as it is not only initial - so in a state of suspense but also if observable completed without any value too. Also keeping rxjs `config` approach can be useful in future if we would like to extend this `config`,
```
export interface FirstValueFromConfig<T> {
  defaultValue: T;
}

export function firstValueFrom<T, D>(source: Observable<T>, config: FirstValueFromConfig<D>): Promise<T | D>;
export function firstValueFrom<T>(source: Observable<T>): Promise<T>;
```

Then again ``` If the source Observable does not complete or emit, you will end up with a Promise that is hung up, and potentially all of the state of an async function hanging out in memory```, 
So they differentiate those two states and do not throw nor return a default value if there is no emission before read.

Initial value could also be achieved by using `startWith` operator but that would not work if the subsequent emission would be empty.

pkozlowski-opensource replied:
> it's mainly useful when you have an Observable<Something> and you want null or undefined as default value.

But this is strange since having `null` as a default value means that anyone can read this value so it should be legitimate type of a signal. In this case it would be `T | null`

kbrilla replied:
Defaulting to `null | undefined`  instead of `initialValue: U` would be ` | async` pipe `| null` problem again

mfp22 replied:
Just throwing in my 2 cents here...

RxJS doesn't distinguish between observables that emit immediately vs asynchronously. So I don't think consumers of observables should rely on knowledge about when they're going to emit. This could cause unexpected errors. Much better would be to type it as `T | null` always, but optimally let devs pass in a type argument (or infer from initial state passed in) if they happen to know 100% that the observable emits immediately. But in my opinion even this explicit type parameter is dangerous because you could modify the observable upstream and TypeScript could not prevent a runtime error. And I don't see a significant difference between `null` and `undefined` here, but I haven't thought that through carefully. 

The only way safe way out of this TypeScript awkwardness is if the observables themselves contained knowledge about whether they are going to emit immediately or not.

muhamedkarajic replied:
I dont like that approach of T | null which would mean I will have always to check for null. Rather then have 2 types one is T other is T | null as it is with/similar to ReplaySubject and BehaviorSubject.

pkozlowski-opensource commented on packages/rxjs-interop/src/from_observable.ts:
```diff
@@ -0,0 +1,90 @@
+
+  // TODO(alxhub): subscription cleanup logic
+
+  return computed(() => {
```
Why is it computed? I guess there are 2 things we need to do:
* make sure that it is not mutable;
* it is branded (?)

Normally a simple function would be enough, I guess this is about branding?

muhamedkarajic replied:
What dose it mean is it branded?

e-oz replied:
> What dose it mean 'branded'?

https://twitter.com/mattpocockuk/status/1625173884885401600

pkozlowski-opensource reviewed (commented):
I just did went over this PR and it LGTM, overall. 

There is obviously remaining work (actually using `DestroyRef` to unsubscribe from observables). 

Other than this I've noted possible discussion topics:
* different type for signal's and initial value
* usage of computed vs. getters

I'm assuming that we've firmly decided on the following:
* `fromObservable` subscribes eagerly, we are not trying to lazy-subscribe;
* we throw when the initial value is not provided (as opposed to defaulting to `null` / `undefined`.

marcincichocki commented on packages/rxjs-interop/test/from_signal_spec.ts:
```diff
@@ -0,0 +1,184 @@
+    // Calling `toPromise` will trigger subscription.
+    const cachedValues = cachedValues$.toPromise();
+
+    // This will resolve `complete` which will also resolve `cachedValues`.
+    await complete;
```
Could you please explain why this is needed? Promises are eager therefore `from(complete)` emits when `cachedValues$` is subscribed. 

I suppose you wanted to create a new promise and save the resolve function, which would be called here to resolve the promise.

muhamedkarajic commented on packages/rxjs-interop/src/from_observable.ts:
```diff
@@ -0,0 +1,90 @@
+  });
+}
+const enum StateKind {
+  NoValue,
```
Is NoValue meaning that there isnt still a value in the observable or is it representing like null | undefined values? I would consider making it 'Loading' state.

BioPhoton commented:
@pkozlowski-opensource I have a question regarding this comment:

> I'm assuming that we've firmly decided on the following:
> 
> * `fromObservable` subscribes eagerly, we are not trying to lazy-subscribe;
> * we throw when the initial value is not provided (as opposed to defaulting to `null` / `undefined`.

Could you please elaborate on the reasoning behind not trying to subscribe lazily?🙏

eneajaho commented:
How is this going to work:

<img width="783" alt="CleanShot 2023-03-14 at 12 49 11@2x" src="https://user-images.githubusercontent.com/25394362/224992114-e097c591-8df2-425a-8eed-70fc58a659b3.png">

The part that throws an error is because, I've implemented the cleanup in a naive way 😬 but with the current apis, that's the only way I see how it can be done.

<img width="976" alt="CleanShot 2023-03-14 at 12 50 09@2x" src="https://user-images.githubusercontent.com/25394362/224992256-e8f7b029-08fc-47f1-9320-7d89e79bd535.png">

mfp22 commented:
I can't think of any reason to define a signal inside an ngOnInit. It throws an actual error just from that code? If so, I think that's a good thing.

eneajaho commented:
> I can't think of any reason to define a signal inside an ngOnInit. It throws an actual error just from that code? If so, I think that's a good thing.

That's not the only issue, we also can't use `fromObservable` outside a component or service, if we depend on the `DestroyRef` that was introduced.

mfp22 commented:
> That's not the only issue, we also can't use `fromObservable` outside a component or service, if we depend on the `DestroyRef` that was introduced.

If there was a way to know when signals had listeners, and subscribe to the observable during that time, then it would also make sense to unsubscribe when the signal ran out of listeners. This would be ideal in my opinion. But as it is, it is conceptually similar to the async pipe. In my opinion it should only be used in components where it will subscribe for exactly as long as the data is needed. But if it could tie the observable's subscription to the signal listeners, that would be ideal. But I'm going to stop commenting on this because I don't really understand the signal mechanics very well and I don't want to crowd this PR. I wish we had comment threads here.

MaximSagan commented:
@alxhub @pkozlowski-opensource Are we targeting v16 for this, or are we still treating this a prototype?
`DestroyRef` [has been merged](https://github.com/angular/angular/commit/0f5c8003ccd1a75516d6a0e31cdb752d031ec430), so we can probably deal with `TODO(alxhub): subscription cleanup logic` now, right?

alxhub committed:
build(core): introduce `@angular/core/rxjs-interop` entrypoint

This commit adds the infrastructure for `@angular/core/rxjs-interop`, a new
core entrypoint which provides bidirectional interoperability between
Angular's built-in reactivity system of synchronous signals, and the RxJS
`Observable` abstraction.

The new entrypoint is set up as an empty shell in this commit, with its
implementation to follow in a followup.

elgreco247 commented on packages/core/src/di/contextual.ts:
```diff
@@ -37,3 +38,22 @@ export function runInInjectionContext<ReturnT>(injector: Injector, fn: () => Ret
+        RuntimeErrorCode.MISSING_INJECTION_CONTEXT,
+        ngDevMode &&
+            (debugFn.name +
+             '() can only be used within an injection context such as a constructor, a factory function, a field initializer, or a function used with `runInInjectionContext`'));
```
Shouldn't the error message in function `injectInjectorOnly` (injector_compatibility.ts) be changed to the one used here?
(mention `runInInjectionContext` instead of `EnvironmentInjector#runInContext`)

eneajaho commented on packages/core/rxjs-interop/src/from_observable.ts:
```diff
@@ -0,0 +1,113 @@
+export function fromObservable<T, U = never>(source: Observable<T>, initialValue?: U): Signal<T|U> {
+  assertInInjectionContext(fromObservable);
+
+  // Note: T is the Observable value type, and U is the intiial value type. They don't have to be
```
```suggestion
  // Note: T is the Observable value type, and U is the initial value type. They don't have to be
```

alxhub committed:
feat(core): implement `takeUntilDestroyed` in rxjs-interop

This commit implements an RxJS operator `takeUntilDestroyed` which
terminates an Observable when the current context (component, directive,
etc) is destroyed. `takeUntilDestroyed` will inject the current `DestroyRef`
if none is provided, or use one provided as an argument.

JeanMeche commented on packages/core/rxjs-interop/src/take_until_destroyed.ts:
```diff
@@ -0,0 +1,34 @@
+ *     passed explicitly to use `takeUntilDestroyed` outside of an injection context. Otherwise, the
+ * current `DestroyRef` is injected.
+ */
+export function takeUntilDestroyed<T>(destroyRef?: DestroyRef): MonoTypeOperatorFunction<T> {
```
What about making this function available in the public API ?

alxhub replied:
Nice catch!

robert-king commented:
I worry a bit about try catch and errors being ignored in some situations, I wonder if something like this could be useful:

```
const counter = fromObservable(counter$);

       effect(() => {
         try {
           seenValue = counter();
           seenError = null;
         } catch (err) {
           seenValue = 0;
           seenError = err;
         }
       });
```
I could imagine sometimes you might want something like this:
```
safeCounter = safeFromObsercable(counter$)
effect(() => {
           [seenValue, seenError] = safeCounter();
})

```

pkozlowski-opensource commented on goldens/public-api/core/rxjs-interop/index.md:
```diff
@@ -4,6 +4,25 @@
+export function fromObservable<T>(obs$: Observable<T>): Signal<T>;
+
+// @public
+export function fromObservable<T, U>(source: Observable<T>, initialValue: U): Signal<T | U>;
```
nit: we should probably align name of the observable, currently we've got `obs$` and `source`.

pkozlowski-opensource commented on packages/core/src/signals/src/computed.ts:
```diff
@@ -6,7 +6,7 @@
  */
 
-import {createSignalFromFunction, defaultEquals, Signal, ValueEqualityFn} from './api';
+import {createSignalFromFunction, DeepReadonly, defaultEquals, Signal, ValueEqualityFn} from './api';
```
nit: unused import?

pkozlowski-opensource committed:
refactor(core): remove DeepReadonly type wrapper for signals

We've been experimenting with the DeepReadonly type that would make
signal values deeply read-only and prevent accidental changes without
going to the owner of data. What we've found out during the experiments
is that additional safety net has more drawbacks than benefits: it just
introduces too much friction to be practical for daily usage.

pkozlowski-opensource commented on packages/core/rxjs-interop/src/from_observable.ts:
```diff
@@ -0,0 +1,115 @@
+      case StateKind.Error:
+        throw current.error;
+      case StateKind.NoValue:
+        throw new Error(`fromObservable() signal read before the Observable emitted`);
```
nit: should it be RuntimeError?

pkozlowski-opensource commented on packages/core/rxjs-interop/test/from_observable_spec.ts:
```diff
@@ -0,0 +1,125 @@
+       effect(() => {
+         seenValue = counter();
+       });
+       await Promise.resolve();
```
thinking out loud: those tests will fail when we change the timing of effects, right? Which kind of rises the question of testing patterns for effects. A separate discussion that we should have.

pkozlowski-opensource replied:
As an alternative we could have written those tests with just a signal and / or computed and just assert that new value is available. We might leave a test or two with `effect` but IMO ones based on just signals would be more readable.

pkozlowski-opensource reviewed (approved):
LGTM

Left a set of bunch of nit / "thinking out loud" comments but nothing that would block this PR. We will need to coordinate this PR landing with other in-flight ones as we will have conflicts around DeepReadonly and timing of effects.

Reviewed-for: public-api 
Reviewed-for: fw-core

jessicajaniuk reviewed (commented):
LGTM other than a non-blocking nit

reviewed-for: fw-core, public-api

jessicajaniuk reviewed (approved):
LGTM other than a non-blocking nit

reviewed-for: fw-core, public-api

e-oz commented:
> I'm assuming that we've firmly decided on the following:
> 
> * we throw when the initial value is not provided (as opposed to defaulting to `null` / `undefined`.

If you already decided on it, why are you discussing it in RFC?
